File: src/main/java/model/organization/event/PolicyEvent.java
package model.organization.event;

import static model.organization.validation.Checks.checkNotNull;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;

import model.organization.entity.Policy;
import model.organization.id.UniqueId;

import com.google.inject.assistedinject.Assisted;

/**
 * The {@linkplain PolicyEvent} class indicates that there is an update about a {@linkplain Policy} entity.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public class PolicyEvent extends AbstractEvent {

	private static final long serialVersionUID = 1100497597698058236L;
	private final UniqueId<Policy> id;
	private transient Integer hashCode = null;
	private transient String toString = null;

	@Inject
	PolicyEvent(@NotNull @Assisted final EventCategory category, @NotNull @Assisted final Policy policy) {
		super(category);
		checkNotNull(policy, "policy");
		id = policy.getId();
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Policy}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Policy> getId() {
		return id;
	}

	@Override
	public boolean equals(final Object object) {
		if (object instanceof PolicyEvent) {
			final PolicyEvent event = (PolicyEvent) object;
			return super.equals(event) && getId().equals(event.getId());
		}
		return false;
	}

	@Override
	public int hashCode() {
		if (hashCode == null) {
			hashCode = super.hashCode() << 16 | getId().hashCode();
		}
		return hashCode;
	}

	@Override
	public String toString() {
		if (toString == null) {
			toString = String.format("%s(%s, %s)", getClass().getSimpleName(), getCategory(), getId());
		}
		return toString;
	}

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/PolicyEvent.kt
package runtimemodels.chazm.model.entity

import runtimemodels.chazm.api.entity.Policy
import runtimemodels.chazm.api.entity.PolicyId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.model.event.AbstractEvent
import runtimemodels.chazm.model.event.EventType
import runtimemodels.chazm.model.message.M
import java.util.*
import javax.inject.Inject

/**
 * The [PolicyEvent] class indicates that there is an update about a [Policy] entity.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
open class PolicyEvent @Inject internal constructor(
    category: EventType,
    policy: Policy
) : AbstractEvent(category) {
    /**
     * Returns a [UniqueId] that represents a [Policy].
     *
     * @return a [UniqueId].
     */
    val id: PolicyId = policy.id

    override fun equals(other: Any?): Boolean {
        if (other is PolicyEvent) {
            return super.equals(other) && id == other.id
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(category, id)

    override fun toString(): String = M.EVENT_WITH_1_ID[super.toString(), id]

    companion object {
        private const val serialVersionUID = 1100497597698058236L
    }
}
